### PR TITLE
DAVAI-126: In-browser Local models — WebLLM engine + focused-tools architecture

### PR DESCRIPTION
**Pull request by ddamelin**
## Summary

Adds an in-browser **Local** model path to DAVAI: Qwen3 1.7B/4B running entirely in the user's browser via WebLLM (WebGPU), with no server round-trips and no data leaving the machine. Because small models proved unreliable at composing raw CODAP API requests, the local path is built on a **focused-tools architecture** — 12 purpose-shaped tools with name resolution, corrective errors, and client-computed "graph sketches" — so every number the model speaks is grounded by construction.

Status note: this began as an exploration (branch name reflects that). The pieces valuable to the server path regardless of this PR's fate are tracked separately: DAVAI-129 (batch), DAVAI-130 (caseFormulaSearch doc), DAVAI-131 (graph-sketch port).

## What's included

- **WebLLM engine** in a Web Worker, WebGPU-gated model entries (Qwen3-1.7B / 4B, q4f16), download-progress announcements; weights (~1.1 GB) fetched from HuggingFace on first use and cached.
- **Focused-tools registry** (local path only): get_graph_info, get_case_values, get_stats, find_cases, select_cases, create_graph, update_graph, create_attribute, update_attribute, create_adornment, group_by, sonify_graph. Validation resolves names via a ladder (exact → unique-normalized repair → corrective listing real options; ids accepted; axis/plot-type phrasing like "the Height dot plot" resolves; true duplicates pick newest). Correctives instruct an immediate retry with a paste-ready argument.
- **Graph sketch**: client-computed description facts injected into graph context — quartile clusters, IQR/residual outliers with coordinates, Pearson r (strength word matches the displayed value), LSRL equation + R² in plain words, units when defined, and (for categorical graphs) category counts, largest crosstab combinations, and empty combinations. Kills the invented-categories/units hallucination class; regression-guarded in the eval.
- **Turn loop**: one-JSON-envelope protocol with repeat-call guard, round cap, cancel epochs, 300s watchdog, unclosed-<think> stripping, and malformed-final-envelope recovery (raw JSON is never spoken).
- **Prompt v2**: generated tool docs + datasets schema digest + selected-graph seed, 8K budget with deterministic trim priority (raw values first; sketch survives).
- **Effort toggle** for local models (none|think, default none — measured: think cost ~10× latency with no accuracy gain on the battery).
- **Live eval harness** (dev options): 18-case battery against a real CODAP document with per-case duration, full tool-result traces, incremental console markers; battery order doubles as cross-tool regression tests (hierarchy mutations, selection changes).
- **Shared fixes**: sonification auto-select parity when the local model creates a graph; Begin/Completed response-time debug parity; LSRL adornment reader in getGraphAdornments; jest ignore for .claude/ worktrees; sam-server doc corrections (percentile 0–1 per DAVAI-128, names-first graph examples).

## Evidence

- 838 unit tests / 54 suites, tsc and strict lint clean.
- Live matrix (16-case battery, final runs): **4B/none 16/16 (~231s)**, **1.7B/none 16/16 (~117s)**; categorical-describe fix verified live (18-case battery adds the anti-hallucination cases).
- Score trajectory across the effort: 8/11 with raw create_request → 16/16 with focused tools.
- Every task went through an implement → independent-review gate; the final whole-branch review returned 0 must-fix findings. Known minors are batched in DAVAI-129.

## Reviewer notes

- The server-model path is functionally untouched (its create_request flow is unchanged); shared edits are limited to the items listed above.
- Local models require WebGPU (Chrome/Edge); the entries are hidden/gated elsewhere. Mock mode unaffected.
- Console errors during graph create/update via the DI API are upstream CODAP v3 issues, filed with attributed stacks as CODAP-1442 / CODAP-1443 (operations succeed; noise only).
- Quick try: `?mode=development` → select a Local model (first load downloads weights) → chat, or run the eval battery from Developer Options with the Mammals sample open and a Height dot plot present.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/utils/local-llm/local-llm-service.ts [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+      // rejected after being superseded must stay silent.
+      if (generation === loadGeneration) {
+        engine = null;
+        loadedModelId = null;
+        setState({ status: "error", modelId, error: err instanceof Error ? err.message : String(err) });
+      }
+      throw err;
+    }).finally(() => {
+      if (generation === loadGeneration) loadPromise = null;
+    });
```
[the remark this answers is not in the document]
Comment (ddamelin) [reply src/utils/local-llm/local-llm-service.ts:173]: Good catch — fixed in 7ab3b5a. On a load that rejects while still current (e.g. `CreateWebWorkerMLCEngine` throws before doLoad clears its registration), the `.catch` in `loadEngine` now terminates the worker and nulls `inflightLoad` when it belongs to this generation, so a failed load no longer leaks a live Worker/GPU/WASM until a later load happens to supersede it. Guarded on the generation so it only ever tears down its own worker. RED-verified test added ("terminates the worker of a still-current load that fails").

File: src/utils/local-llm/local-llm-service.ts [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+    // rejection inside the worker — it never reaches this promise, so the turn just hangs
+    // forever. The envelope parser + one-retry loop in local-llm-loop.ts were designed to
+    // survive unconstrained model output, so run unconstrained here. A VERIFIED schema/EBNF
+    // grammar constraint (once 0.2.84's bug is worked around or a fixed version is available)
+    // is a named follow-up, not done here.
+    const create = engine.chat.completions.create({
+      messages,
+      temperature: 0,
+      max_tokens: opts?.maxTokens ?? 1024,
+    });
```
[the remark this answers is not in the document]
Comment (ddamelin) [reply src/utils/local-llm/local-llm-service.ts:197]: Fixed in 7ab3b5a — attached a no-op `create.catch(() => undefined)` so a late rejection after the timeout wins the race is always observed rather than becoming an unhandled rejection. It runs independently of the `Promise.race` (which registers its own handlers), mirroring the existing guard on `loadEngine`'s `settling` promise just above. Note: I did not add a dedicated unit test for this one — Node's `unhandledRejection` event doesn't fire deterministically within jest's fake-timer microtask flushing (a listener-based test passed even with the guard removed, i.e. no teeth), so rather than ship a false-confidence test I'm relying on parity with the already-tested `loadEngine` guard and the existing timeout test. Open to a suggestion if you have a reliable pattern for it.

File: src/utils/local-llm/local-llm-envelope.ts [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+  const trimmed = s.endsWith("\\") && !s.endsWith("\\\\") ? s.slice(0, -1) : s;
+  return trimmed.replace(/\\(.)/g, (_match, ch) => {
+    switch (ch) {
+      case "n": return "\n";
+      case "t": return "\t";
+      case "r": return "\r";
+      default: return ch; // \" \\ \/ and anything else: the escaped character itself
+    }
+  });
+};
```
[the remark this answers is not in the document]
Comment (ddamelin) [reply src/utils/local-llm/local-llm-envelope.ts:93]: Fixed in 7ab3b5a — `unescapeJsonString` now decodes via `JSON.parse('"' + trimmed + '"')` first, which correctly handles every standard JSON escape including `\uXXXX`, `\b`, and `\f`; the hand-rolled switch (now also covering `\u`/`\b`/`\f`) remains as a fallback for the rare capture JSON.parse still rejects (a raw control char or stray escape the tolerant regex admitted). This matters for international users (cf. CODAP-1440's Chinese attribute names) — a truncated "café" now recovers correctly instead of speaking "cafu00e9". RED-verified test added with `\u00e9`-escaped accented text.

File: src/components/App.test.tsx [hunk 4]
```diff
@@ -162,6 +205,32 @@ describe("test load app", () => {
...
+
+    fireEvent.change(screen.getByTestId("chat-input-textarea"), { target: { value: "describe the graph" } });
+    fireEvent.click(screen.getByTestId("chat-input-send"));
+
+    expect(mockAssistantStore.handleMessageSubmitLocalLlm).toHaveBeenCalledWith("describe the graph");
+    expect(mockAssistantStore.handleMessageSubmit).not.toHaveBeenCalled();
+  });
+
+  it("sets the effort before dispatching a local submission, mirroring the server branch " +
+    "(DAVAI-126 thinking toggle)", () => {
```
Comment (emcelroy) [inline src/components/App.test.tsx]: This concatenation is a bit odd. I would just drop the `+
    "(DAVAI-126 thinking toggle)"` part. I don't think recording the Jira story ID is necessary.

File: src/components/App.test.tsx [hunk 5]
```diff
@@ -96,10 +116,33 @@ describe("test load app", () => {
...
+      return mockUnsubscribeLoadStateChange;
+    });
   });
 
   afterEach(() => {
     cleanupMockSpeechSynthesis();
+    // Several assistantStore/CODAP mocks are shared module-level jest.fn()s; clear their
+    // call history between tests so a submit assertion here can't see a call left over
+    // from an earlier test in this file (e.g. handleMessageSubmit from a prior submit).
+    jest.clearAllMocks();
```
Comment (emcelroy) [inline src/components/App.test.tsx]: This is kind of a nitpick, sorry, but AI tends to leave a lot of comments like this that I don't think are necessary and just contribute to the code getting "noisy". Calling `clearAllMocks()` in an `afterEach` is a common pattern that really shouldn't need explaining.

File: src/components/App.tsx [hunk 6]
```diff
@@ -154,8 +156,92 @@ export const App = observer(() => {
   useEffect(() => {
     // Initialize the assistant on mount and when the LLM ID changes.
     handleInitializeAssistant();
+
+    // Drive the local in-browser engine's lifecycle off the same LLM-selection change:
+    // load it when a Local model is selected, and unload it (freeing GPU/WASM memory)
+    // when switching away to a server-backed model.
+    const entry = findEntryByLlmId(appConfig.llmList as any, appConfig.llmId);
```
Comment (emcelroy) [inline src/components/App.tsx]: This cast to `any` seems unnecessary.

```suggestion
    const entry = findEntryByLlmId(appConfig.llmList, appConfig.llmId);
```

File: src/components/App.tsx [hunk 7]
```diff
@@ -154,8 +156,92 @@ export const App = observer(() => {
...
+      } else if (s.status === "ready") {
+        lastMilestone = 0;
+        milestoneModelId = undefined;
+        announcedFirstProgress = false;
+        transcriptStore.addMessage(DAVAI_SPEAKER, { content: "The local model is ready.", kind: "announcement" });
+      }
+    });
+    return off;
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
```
Comment (emcelroy) [inline src/components/App.tsx:217]: This works fine as is, so you can leave it if you want, but my inclination would be to move the load-announcement logic to a separate helper function or hook. That would help declutter the `App` component and also the App.test.tsx tests since you could test the logic separately without having to render the App component.

File: src/utils/local-llm/graph-sketch.ts [hunk 8]
```diff
@@ -0,0 +1,507 @@
...
+
+const pearsonR = (xs: number[], ys: number[]): number => {
+  const n = xs.length;
+  const mx = xs.reduce((a, b) => a + b, 0) / n;
+  const my = ys.reduce((a, b) => a + b, 0) / n;
+  const cov = xs.reduce((sum, x, i) => sum + (x - mx) * (ys[i] - my), 0);
+  const sdx = Math.sqrt(xs.reduce((sum, x) => sum + (x - mx) ** 2, 0));
+  const sdy = Math.sqrt(ys.reduce((sum, y) => sum + (y - my) ** 2, 0));
+  return cov / (sdx * sdy);
+};
```
Comment (emcelroy) [inline src/utils/local-llm/graph-sketch.ts:121]: Claude caught this which seems worth addressing:

When either axis has zero variance, `pearsonR` returns NaN; the current comparisons classify that as positive, strong (r = NaN), and the fallback R² is also NaN. Detect a zero denominator and emit an “undefined correlation” description instead.

Also applies to: 270-285

File: src/utils/local-llm/local-llm-prompt.ts [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+  while (total(out) > maxChars && out.length > 2) out.splice(1, 1);
+  // 3. Truncate the schema digest section tail.
+  if (total(out) > maxChars) {
+    const others = total(out.slice(1));
+    const room = Math.max(0, maxChars - others - 64);
+    const content = sys().content;
+    const keep = content.slice(0, room);
+    const trailingSwitch = THINK_SWITCHES.find((s) => content.endsWith(s));
+    const switchSuffix = trailingSwitch ? `\n${trailingSwitch}` : "";
+    out[0] = { ...sys(), content: `${keep}\n[schema digest truncated]${switchSuffix}` };
```
Comment (emcelroy) [inline src/utils/local-llm/local-llm-prompt.ts]: Claude flagged this. I'm not sure how worried about this you may be, but my understanding is a large current user message could significantly shrink `room`, and when `room` drops below the length of the instructions plus tool documentation, the slice would cut through the tool documentation. (One nitpick I'd add is it'd be good to have a helpfully-named constant for the magic number on line 74.)

Preserve instructions and tool documentation during trimming.

`content.slice(0, room)` truncates the entire system prompt, so a large schema digest can cut through the tool definitions placed earlier by `buildLocalSystemPrompt`. The model may then lose the capability surface and stop making tool calls. Truncate only the schema-digest section while preserving instructions, tool docs, the graph structure, and the switch token.

File: src/constants.ts [hunk 10]
```diff
@@ -9,6 +9,12 @@ export const STREAMING_STATUS = "streaming";
 
 export const GREETING = `Hello! I'm DAVAI, your Data Analysis through Voice and Artificial Intelligence partner.`;
 
+// Shown when a Local model is selected but the browser lacks WebGPU. Shared by App.tsx (engine
+// lifecycle) and assistant-model.ts (submit-time guard) so the two paths speak identically.
```
Comment (emcelroy) [inline src/constants.ts]: Here's another example of the kind of comment that doesn't seem necessary and adds noise, in my opinion. The name of the constant implies when/why it would be used, and the fact that it's defined and exported from `constants.ts` implies it's used in multiple places.

**Review by emcelroy — CHANGES_REQUESTED** [review]
Generally looking good 👍 The local LLMs worked pretty well when I tried them in the browser. 

Claude flagged a bunch of issues, at least a couple of which seem significant enough to block merging. I also noticed some other small things that could be changed. See the main comment below and the few inline comments I left elsewhere.

The following was developed with Claude Code.

### Overall

Large, ambitious PR (~11k lines) adding a complete WebLLM-based in-browser inference path: a worker-hosted engine with a load-state machine, an agentic tool loop, an envelope parser, a CODAP tool suite, and a graph "sketch" generator running client-side alongside the server path.

The engineering quality is high. The load-state machine (generation counter, superseded-load abandonment, late-rejection silencing, worker-leak-on-failure) is careful and well-tested; the envelope parser never throws and always degrades; MST action/flow discipline is respected; test coverage is broad. No stability defects (crashes/hangs/data-loss) surfaced — the issues are about **correctness of what gets said**.

That's the systemic theme: because this output is spoken aloud to blind/low-vision users, **"confidently wrong" text is the dominant risk class**. A `NaN`, an `"n/a"`, a mislabeled direction, or a raw JSON blob is the tool authoritatively asserting something false to a user who can't see the graph to catch it. Most medium findings are instances of that pattern.

Recommendation: **approve with changes** — two gating items, plus a cluster of confident-wrong-speech bugs to fix.

### Gate on these

**1. Prompt trimming can sever the tool docs → model silently stops calling tools.** `local-llm-prompt.ts` step 3 does a blind `content.slice(0, room)` over the whole system prompt. `room` shrinks with the size of the current user message, and nothing anchors the cut to the schema-digest boundary. A large user turn drives `room` below `len(instructions + toolDocs)` and the slice eats the tool documentation — the "instructions and tool docs are never trimmed" comment isn't enforced by the code. It also cuts the graph structure that step 1 deliberately preserved, and always labels the result `[schema digest truncated]` regardless of what was removed. Plausible trigger, silent, disables the core capability. Fix: anchor truncation to the `DIGEST_HEADER` offset, clamp so the cut never precedes the digest, add tests for a large user message and a large digest.

**2. `resolveGraph` throws on a numeric `graph` id — the case it was built to accept.** `resolve.ts:369`. Callers pass `args.graph as string` (compile-time-only cast); when the model echoes a printed id as an unquoted JSON number, `args.graph` is a runtime `number`, passes the `undefined/""` guard, and `.trim()` throws `TypeError`. `dispatchTool` surfaces this as `... failed with an internal error: requested.trim is not a function` — misleading and non-corrective, for exactly the numeric-id-echo path the rung-0 comment describes. Fix: coerce `String(requested)` at the top of `resolveGraph`; add a test with a `number`-typed arg.

### Should-fix before real users (confident-wrong speech)

**3. Constant / zero-variance axis → "positive, strong (r = NaN)."** `pearsonR` returns `0/0 = NaN`; `directionWord`→"positive", `strengthWord`→"strong", and the R² line prints `NaN%`. Detect the zero denominator and replace the whole Relationship clause with an "undefined correlation — every point shares the same X value" sentence; guard the R² line too. Highest-priority speech bug: plausible user graph, authoritative-sounding output.

**4. Scatter outlier label contradicts the values it lists.** `graph-sketch.ts:299-310`. The "Unusually high/low X" word is taken from the top outlier only, but the clause enumerates outliers on both sides of the median — so a low outlier is read out under "unusually high." Label each listed outlier by its own side, or split into separate high/low clauses.

**5. Forced-final loop path can read raw tool-call JSON aloud.** `local-llm-loop.ts:118,141`. If the model answers the "answer now" nudge with another tool envelope, `parseEnvelope` returns `tool_call`, the code falls through to `return stripThink(lastRaw)`, and the user hears `{"tool":"get_stats",...}` spoken out — the very thing `recoverFinalResponse` prevents for final envelopes. Degrade a non-final/invalid forced result to `FALLBACK_RESPONSE`.

**6. `find_cases` reports categorical `where`-matched values as "n/a".** `find-cases.ts:184`. A condition like `Diet = 'meat'` runs numeric `formatValue` on the categorical stat attribute, yielding `Lion (Diet n/a)` — "n/a" reads as *missing* to a listener, about a value that's present. Handle non-numeric stat attributes.

**7. `createThread` (reset) doesn't clear the local message queue.** `assistant-model.ts`. `handleCancel` and `setLlmId` both clear the queue so the `onSnapshot` reactor can't route a queued local message to the server path; `createThread` was missed, so a reset during an in-flight local turn can leak a local message to the server. Latent today (not wired to UI), one-line fix.

**8. `resolveDataContext` has no sole-context default.** `resolve.ts:41`. Unlike `resolveCollection`, it doesn't default to the only dataset, so when the model omits `dataContext` in a single-dataset document it's bounced with `Unknown data context ""` instead of resolving silently. Forces avoidable round-trips.

### Minor / latent / polish (ticket, don't block)

- **More NaN/degenerate formatting:** `roundSig(NaN|±Infinity)` → bare `"NaN"`/`"-NaN"`; `directionWord(0)` → "positive" for an uncorrelated axis; floating-point `r²` slightly >1 → "explains about 101%"; `get-stats` `toPrecision(6)` → `"1.23457e+6"` spoken aloud for large means.
- **Coercion trusting the wrong source:** whitespace-only cells (`Number(" ") === 0`) counted as real zeros in `get-stats`/`find-cases`; `get-case-values` renders blank cells as literal `"null"`/`"undefined"`; `find-cases` label attribute relies on `attr.type === "categorical"` even though `get-stats` documents `type` as unreliable and value-sniffs — make that stance consistent.
- **LSRL data-shape assumption:** LSRL paths read `data[0]` and assume `{slope,intercept,rSquared}`; a legend-split scatter (one line per category) silently reports only the first, and the SD adornment never states the SD value. Verify against a live CODAP scatterplot; degrade gracefully rather than `"slope undefined"` if the shape differs.
- **Selected-cases summary** ("at ." / miscount on non-numeric coords) — real but latent, since nothing populates `selectedPairs` yet.
- **Resolver hardening:** `normalizeName` doesn't re-trim after collapsing separators (a leading `_` breaks the normalized-match repair); rung-3 axis scoring uses unweighted substring containment so a name and its superset tie; no duplicate-name guard in `registerTools`; resolvers rely on callers to `String(...)`-coerce rather than coercing at the boundary (root cause of #2).
- **Loop robustness:** `generate()` rejections (e.g. the watchdog timeout) aren't caught inside `runLocalTurn`, so they escape rather than degrading to `FALLBACK_RESPONSE` — confirm the caller handles it or catch locally; A-B-A repeats of a mutating tool can double-execute (the repeat guard is consecutive-only).
- **Nits:** `skipLibCheck: true` is repo-wide, not scoped to web-llm; the local engine isn't unloaded on App unmount; the `64`-char reserve in prompt trimming is an un-named magic number that under-reserves if the marker string grows; `unescapeJsonString`'s fallback turns a malformed `\u` into a NUL char; missing-name fields render `"undefined"` into prompt digests.

### Test coverage

Strong overall, but the gaps line up with the findings above. Worth adding before merge: numeric-typed `resolveGraph` arg (#2); prompt trim with an oversized user message / digest (#1); constant-axis scatter asserting the spoken Relationship + R² prose (#3); mixed-direction scatter outliers (#4); forced-final returning another tool call (#5); categorical `where`/`orderBy` value display (#6); reset-during-local-turn queue clearing (#7).

### Bottom line

Architecturally sound and unusually well-defended for a PR this size — the runtime/state-machine layer is merge-quality. Hold on the two gating items (#1 prompt-trim tool-doc loss, #2 numeric-id throw), then clear the confident-wrong-speech cluster (#3–#6), which is where the user-facing risk concentrates given the accessibility mission. The rest can be follow-up tickets.

File: src/components/App.test.tsx (hunk 4 above)
Comment (ddamelin) [reply src/components/App.test.tsx]: Done in 08d5e23 — dropped the Jira-story suffix from the test name.

File: src/components/App.test.tsx (hunk 5 above)
Comment (ddamelin) [reply src/components/App.test.tsx]: Agreed — removed in 08d5e23. Fair critique of agent-written code generally; we'll keep an eye on comment noise in the rest of the diff too.

File: src/components/App.tsx (hunk 6 above)
Comment (ddamelin) [reply src/components/App.tsx]: Applied your suggestion in 08d5e23 — the cast was indeed unnecessary; tsc is clean without it.

File: src/components/App.tsx (hunk 7 above)
Comment (ddamelin) [reply src/components/App.tsx:217]: Took you up on it in 08d5e23 — the load-announcement effect body is now `createLoadAnnouncer` (src/components/local-model-load-announcer.ts), a plain stateful helper the effect delegates to, with its own focused unit tests (App.test.tsx no longer has to render the whole component to exercise announcement logic).

File: src/utils/local-llm/graph-sketch.ts (hunk 8 above)
Comment (ddamelin) [reply src/utils/local-llm/graph-sketch.ts:121]: Fixed in 3322850. Zero-variance on either axis now short-circuits pearsonR's NaN path entirely: the Relationship line becomes an explicit "undefined — every point has the same <axis>" sentence and the R²/farthest-from-line lines are suppressed. Also hardened the neighbors while in there: r === 0 now reads "no linear relationship" (directionWord no longer calls 0 "positive"), displayed R² is clamped to ≤100% against float drift, and roundSig never emits "NaN"/"Infinity" as speakable text. Constant-x, constant-y, r=0, and clamp fixtures assert the full spoken lines.

File: src/utils/local-llm/local-llm-prompt.ts (hunk 9 above)
Comment (ddamelin) [reply src/utils/local-llm/local-llm-prompt.ts]: Fixed in 6b3da9a — treated as a gate, you were right to worry: `room` shrinks with the current user message, so a ~14K-char paste (e.g. CSV) reached the tool docs. Truncation is now anchored to the digest header offset — instructions, tool docs, and the graph-structure section are structurally untrimmable (the cut can never precede the digest), and if even that prefix exceeds the budget we keep it whole anyway (slightly overshooting beats silently disabling tool-calling). The magic 64 is now TRIM_MARKER_RESERVE_CHARS, sized from the actual marker strings. Tests cover both triggers: oversized user message and oversized digest, each asserting the never-trimmed prefix is byte-identical, plus an honest truncation label.

File: src/constants.ts (hunk 10 above)
Comment (ddamelin) [reply src/constants.ts]: Removed in 08d5e23.

**Comment by ddamelin** [comment]
Thanks for the thorough review — the systemic framing ("confidently wrong speech is the dominant risk class") is exactly the right lens for this app, and both gates were real. All findings are addressed across four commits (`6b3da9a` gates → `3322850` speech cluster → `08d5e23` one-liners + your inline items → `8a7be12` minor cluster): **897 tests (+57), tsc clean, lint clean, zero drops.**

**Gates:**
- **#1 prompt trim** — you were right, and this was worse than our own review had triaged it (we'd rated the missing enforcement "unreachable at real budget," which ignored that `room` shrinks with the current user message). Truncation is now anchored to the digest header: instructions/tool docs/graph structure are structurally untrimmable, byte-identical under both an oversized user message and an oversized digest (both tested); honest label; `TRIM_MARKER_RESERVE_CHARS` named.
- **#2 numeric id** — fixed at the boundary per your root-cause note: all resolvers coerce `String(requested)` on entry, so the compile-time `as string` fiction can't produce a runtime TypeError anywhere. Number-typed id test added; the closure-property and identifier-consistency integration tests still pass.

**Speech cluster (#3–#6):** all fixed — zero-variance → explicit "undefined correlation" sentence (plus `r === 0` → "no linear relationship", R² clamp, no speakable NaN/Infinity anywhere); per-side outlier labels; forced-final now degrades tool-call/JSON-shaped output to the safe fallback **while still speaking plain-prose answers** (a prose reply to the nudge is a good answer — only JSON shapes are unspeakable); find_cases prints categorical values verbatim and omits the parenthetical for blanks.

**#7 (createThread queue):** confirmed — and it reproduced *worse* than described: under test the leaked queue routing produced a hard crash, not just a mis-route. Fixed + regression test. **#8:** sole-context default added, mirroring `resolveCollection`'s precedent.

**Minor cluster:** all fixed except three, plus one deliberate hold:
- **Held: boolean→1/0 coercion stays** — the mean of a true/false attribute is the proportion true, which is a meaningful statistic; the locking test now documents that as intentional. (Whitespace-→0, which silently corrupts real means, *is* fixed.) Happy to revisit if you disagree.
- **Deferred, with reasons:** `skipLibCheck` scoping (the flag is global-only in TypeScript — scoping requires a different mechanism entirely, its own task); the A-B-A repeat guard (a loop-semantics change we'd want validated against a fresh eval-matrix run, not slipped into a review commit); LSRL multi-legend response shape (needs a live CODAP scatterplot-with-legend to know the real payload — but the graceful-degrade guard is in now, so a shape mismatch omits the line rather than speaking "slope undefined"). All three tracked in DAVAI-129.
- **Answered rather than changed:** `generate()` rejections escaping `runLocalTurn` — the caller handles this deliberately: `handleMessageSubmitLocalLlm` catches, posts the error message, and logs completion timing, so a watchdog timeout surfaces as an honest error rather than a silent fallback.

Inline comments each have a threaded reply with the commit. Re-requesting review.

**Comment by ddamelin** [comment]
One more commit on top of the review response: `e026cf7`. We ran an adversarial second-pass review (Codex) over the five fix commits themselves, and it found six edge cases in the new guards — all confirmed and hardened:

- forced-final now de-fences markdown-wrapped output before the speakability check (a fenced, unclosed tool-call attempt degrades to the safe fallback; fenced *prose* is still spoken, minus the fences);
- `Number.isFinite` replaces `typeof`/nullish checks at the LSRL gate and adornment formatter (a NaN `rSquared` drops only the R² sentence; a NaN slope drops the block; no non-finite value reaches spoken text);
- `roundSig` no longer throws on denormal-scale values or emits e-notation at the ≥1e21 boundary;
- `null` and whitespace-only now count as *omitted* at every resolver default (graph → selected, sole data context, single/leaf collection) — `{"graph": null}` resolves the selected graph instead of "No graph matches \"null\"".

926 tests (+29), tsc/lint clean. Commit `7ab3b5a` came back from that pass with no findings.

**Comment by ddamelin** [comment]
Also landed `19723c5` — a comments-only pass over the branch (53 files, net −184 lines): stripped the agent's round-numbering/review-attribution provenance and change-narration comments, kept the load-bearing rationale (external-constraint workarounds, invariants, a11y reasoning). Directly in response to the comment-noise feedback; reviewable in isolation since it touches no code, strings, or logic — 926/926 tests, tsc and lint clean.

File: src/utils/local-llm/local-llm-loop.ts [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+
+const capToolResult = (result: string): string =>
+  result.length <= MAX_TOOL_RESULT_CHARS
+    ? result
+    : `${result.slice(0, MAX_TOOL_RESULT_CHARS)}\n${TOOL_RESULT_TRUNCATED_MARKER}`;
+
+// Identifies a tool call by name + arguments so an identical repeat can be recognized regardless
+// of key order sensitivity concerns: args come from parseEnvelope's own JSON.parse of the model's
+// output, so two calls with the same semantic args produce the same object shape/key order here.
+const callKey = (name: string, args: Record<string, unknown>): string => `${name}::${JSON.stringify(args)}`;
```
[the remark this answers is not in the document]
Comment (ddamelin) [reply src/utils/local-llm/local-llm-loop.ts]: Good catch — fixed in 1d80b94. `callKey` now sorts object keys at every level before stringifying (a small recursive JSON.stringify replacer), so a semantically-identical repeat with reordered keys is still recognized. Added a regression test that reproduces the exact failure mode you described and RED-verified it against the old bare-stringify behavior (reverted the fix locally, confirmed the test failed, restored it).

File: src/utils/local-llm/tools/update-attribute.ts [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+    const attr = resolveAttribute(String(args.attribute ?? ""), dc.value);
+    if (!attr.ok) return { ok: false, error: attr.error };
+
+    const newName = typeof args.newName === "string" && args.newName.trim() ? args.newName.trim() : undefined;
+    const description = typeof args.description === "string" && args.description ? args.description : undefined;
+    const unit = typeof args.unit === "string" && args.unit ? args.unit : undefined;
+    const formulaRaw = typeof args.formula === "string" && args.formula ? args.formula : undefined;
+    const hasPosition = args.position !== undefined && args.position !== null && args.position !== "";
+    const positionArg = hasPosition ? Number(args.position) : undefined;
+    const position = positionArg !== undefined && Number.isFinite(positionArg) ? Math.floor(positionArg) : undefined;
```
[the remark this answers is not in the document]
Comment (ddamelin) [reply src/utils/local-llm/tools/update-attribute.ts:22]: Fixed in 1d80b94 — validate() now rejects a negative position (including the `Math.floor(-0.1) === -1` case specifically, which a naive pre-floor check would miss) rather than forwarding it to CODAP. Two-part regression test covers both the whole and fractional negative inputs.

File: src/utils/local-llm/tools/update-graph.test.ts [hunk 13]
```diff
@@ -0,0 +1,247 @@
...
+    expect(send).not.toHaveBeenCalled();
+  });
+
+  it("rejects an unknown legendAttribute before any request", () => {
+    const v = updateGraphTool.validate({ graph: "Height vs Mass", legendAttribute: "Speed" }, ctx);
+    expect(v.ok).toBe(false);
+    expect(send).not.toHaveBeenCalled();
+  });
+
+  it("rejects an unknown title-only change is fine — title needs no resolution", () => {
```
[the remark this answers is not in the document]
Comment (ddamelin) [reply src/utils/local-llm/tools/update-graph.test.ts]: Fixed in 1d80b94 — renamed to "accepts a title-only change — title needs no resolution", matching what it actually asserts (`v.ok === true`). Looks like an artifact of an edit that merged two test descriptions into one.

**Comment by ddamelin** [comment]
Addressed Copilot's latest pass in `1d80b94`: 928 tests (+2), tsc/lint clean.
- repeat-call guard now sorts arg keys before comparing, so a reordered-but-identical retry is still caught (previously could double-execute a mutating tool)
- `update_attribute`'s `position` rejects negative values, including the floor(-0.1)=-1 edge case
- fixed a mislabeled test name in update-graph.test.ts (said "rejects", asserted acceptance)

All three verified against the actual code before fixing; threaded replies on each.

**Review by emcelroy — APPROVED** [review]
Looks good 👍

**Comment by ddamelin** [comment]
One more cleanup pass in `69f167c`: the earlier comments-only pass (`19723c5`/`25c7f8b`) deliberately left test-name strings alone per its own scope rule — this one extends the same standard to `it()`/`describe()` descriptions. Stripped ~121 Jira/PR-review/task tags like "(DAVAI-126 Task 11)" and "(PR #114 review item 10)" from test names across 27 files, plus a couple of comments and a console.log the last fix commit had reintroduced. A full repo-wide scan now returns zero remaining references. 928/928 tests unchanged (no renames added/removed/duplicated any test), tsc and lint clean.
